### ISSUE TITLE
Are the plans for UPnP/DLNA support?

### ISSUE THREAD

**Issue by yfr** [mafintosh/playback#20]
I would love an easy to use obvious way to use our local network streaming. Any plans anyone?

**Comment by mafintosh** [comment]
are there any good node modules available for adding this?

**Comment by yfr** [comment]
I don't really know. Just popped in mind that this would be cool. Will have a look, but i'm not sure if i can be a helping hand at this.

**Comment by SchizoDuckie** [comment]
There are some upnp/dlna modules in the popcorntime repos that may help you.

**Comment by freeall** [comment]
@SchizoDuckie nice. Do you have a link?

**Comment by thibauts** [comment]
I was thinking about it just this morning :) I have this https://github.com/thibauts/node-upnp-mediarenderer-client

As far as I know popcorntime uses my module. I can't (obviously) test it on many TVs  but it seems to work with a good deal of brands.

**Comment by thibauts** [comment]
What's missing : 

- Maybe a clean module for SSDP. I have code sleeping somewhere and could publish it. But if you think what's on NPM is enought I won't bother.
- Maybe UPnP events to be able to sync with the TV remote actions like pause, FF, etc. This could be implemented here https://github.com/thibauts/node-upnp-device-client

Regarding external subtitles support :
- Subtitles seem to work fine on Samsung (and probably others). They're advertised in the HTTP server's response headers when serving the media file. See this gist https://gist.github.com/thibauts/5f5f8d8ce6566c8289e6
- Other brands (like my Sony) advertise them in a media description (I have to refresh my memory on this) using the DIDL-Lite format. Open issue here https://github.com/thibauts/node-upnp-mediarenderer-client/issues/6

I think that's pretty much it.

It has to be noted that good feedback seems to be difficult to collect, given the large variety of brands and embedded UPnP engines. I've seen the popcorn time dev struggle with this, as users tended to mistake formats unspported by their TV with a bug. There's a function [here](https://github.com/thibauts/node-upnp-mediarenderer-client/blob/master/index.js#L11) that should help detect unsupported formats before trying to play them.

**Comment by thibauts** [comment]
Currently working on the eventing part.

**Comment by SchizoDuckie** [comment]
This would be awesome if i could integrate with DuckieTV.

I've been putting off chromecast development to await vlc's implementation,
but if your project also starts working on Linux/windows and can support
launching via a custom url protocol it would be a match made in heaven

**Comment by thibauts** [comment]
I just finished implementing eventing in [upnp-device-client](https://github.com/thibauts/node-upnp-device-client) and added debug traces. I'll try to find some time to work on media events in upnp-mediarenderer-client next week.

**Comment by thibauts** [comment]
Looks like it wasn't THAT much work :) media events implemented in [upnp-mediarenderer-client](https://github.com/thibauts/node-upnp-mediarenderer-client).

**Comment by xat** [comment]
@thibauts amazing work!
Playing around with it right now :D

**Comment by thibauts** [comment]
Just updated [upnp-mediarenderer-client](https://github.com/thibauts/node-upnp-mediarenderer-client) with DIDL-Lite subtitles support. Some TVs don't support external subtitles at all, though (mine included). So testers welcome !

**Comment by thibauts** [comment]
So what we need now is some module for discovery. `node-ssdp` would do the trick but I don't feel it, I don't really know why. If someone is interested in this feature, let me know what you think.

**Comment by Glavin001** [comment]
I would really love this feature! Is there any development on this lately?

**Comment by thibauts** [comment]
I left this aside as there doesn't seem to be much interest in this feature. Though the UPnP module works and is ready to be integrated.

As for subtitles, I'm not sure it would be that easy with streams :/

**Comment by Glavin001** [comment]
I'd love to help this weekend / next week and get this integrated!

Looking at the [usage](https://github.com/thibauts/node-upnp-mediarenderer-client#usage)

```javascript
// Instanciate a client with a device description URL (discovered by SSDP)
var client = new MediaRendererClient('http://192.168.1.50:4873/foo.xml');
```

Is there an example of what `foo.xml` is?

I've found this package that uses your library and the section of code that generates their XML: https://github.com/gyzerok/nodecast-js/blob/master/src/Browser.js#L10-L29
It'd be nice to have a clear example though.

**Comment by thibauts** [comment]
Yeah what's provided to MRC is the URL that's in the `Location` header of the SSDP reply.

Having an example would be nice, sure. I didn't do it so far as I wasn't very satisfied with the available options regarding SSDP modules.

**Comment by thibauts** [comment]
`foo.xml` is the root URL of the UPnP device. It's the entry point from which you can introspect the device and find services / capabilities. As said above it's also the URL that the device announces on the local network through SSDP.
